[docs][coroutines] Revamp "Debugging C++ coroutines" (#142651)
llvm/llvm-project: clang/docs/DebuggingCoroutines.rst

Introduction
============

For performance and other architectural reasons, the C++ Coroutines feature in
the Clang compiler is implemented in two parts of the compiler.  Semantic
analysis is performed in Clang, and Coroutine construction and optimization

> takes place in the LLVM middle-end.

Author: Agree, it makes sense to keep this information around.

I would move it to the "Toolchain Implementation Details", though, i.e. further down.

I don't think having this already in the introductory section makes sense.
I think most clang / coroutine users primarily care about what they can do and how they can do so when debugging coroutines.
The reasons on why certain things don't work are imporant for us toolchain developes, but probably not that relevant to most end users.

> takes place in the LLVM middle-end.

Author: I merged this now, since I want to be able to point people to the documentation from a discourse thread. Let me know if I missed anything. Happy to update this further also post-commit

> Coroutines in C++ were introduced in C++20, and their user experience for

Reviewer: ```suggestion
Coroutines in C++ were introduced in C++20, and the user experience for
```

Coroutines are generally used either as generators or for asynchronous
programming. In this document, we will discuss both use cases. Even if you are
using coroutines for asynchronous programming, you should still read the
generators section, as it will introduce foundational debugging techniques also
applicable to the debugging of asynchronous programming.

Both compilers (clang, gcc, ...) and debuggers (lldb, gdb, ...) are

> still improving their support for coroutines. As such, we recommend to use the

Reviewer: ```suggestion
still improving their support for coroutines. As such, we recommend using the
```

latest available version of your toolchain.

This document focuses on clang and lldb. The screenshots show
[lldb-dap](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.lldb-dap)
in combination with VS Code. The same techniques can also be used in other
IDEs.

Debugging clang-compiled binaries with gdb is possible, but requires more
scripting. This guide comes with a basic GDB script for coroutine debugging.

This guide will first showcase the more polished, bleeding-edge experience, but
will also show you how to debug coroutines with older toolchains. In general,
the older your toolchain, the deeper you will have to dive into the
implementation details of coroutines (such as their ABI). The further down in

> this document, the more low-level, technical the content will become. If you

Reviewer: ```suggestion
this document you go, the more low-level, technical the content will become. If you
```

you are on an up-to-date toolchain, you will hopefully be able to stop reading
earlier.

Debugging generators
====================

> The first major use case for coroutines in C++ are generators, i.e., functions

Reviewer 2: it sounds better

which can produce values via ``co_yield``. Values are produced lazily,
on-demand. For that purpose, every time a new value is requested the coroutine
gets resumed. As soon as it reaches a ``co_yield`` and thereby returns the
requested value, the coroutine is suspended again.

This logic is encapsulated in a ``generator`` type similar to

.. code-block:: c++

  // generator.hpp
  #include <coroutine>

  // `generator` is a stripped down, minimal generator type.
  template<typename T>
  struct generator {
    struct promise_type {
      T current_value{};

      auto get_return_object() {
        return std::coroutine_handle<promise_type>::from_promise(*this);
      }
      auto initial_suspend() { return std::suspend_always(); }
      auto final_suspend() noexcept { return std::suspend_always(); }
      auto return_void() { return std::suspend_always(); }
      void unhandled_exception() { __builtin_unreachable(); }
      auto yield_value(T v) {
        current_value = v;
        return std::suspend_always();
      }
    };

    generator(std::coroutine_handle<promise_type> h) : hdl(h) { hdl.resume(); }
    ~generator() { hdl.destroy(); }

>     generator<int>& operator++() { hdl.resume(); return *this; } // resume the coroutine
>     int operator*() const { return hdl.promise().current_value; }

Reviewer: ```suggestion
    generator<T>& operator++() { hdl.resume(); return *this; } // resume the coroutine
    T operator*() const { return hdl.promise().current_value; }
```

a regular function.

.. image:: ./coro-generator-variables.png

Note the two additional variables ``__promise`` and ``__coro_frame``. Those
show the internal state of the coroutine. They are not relevant for our
generator example, but will be relevant for asynchronous programming described
in the next section.

Stepping out of a coroutine
---------------------------

When single-stepping, you will notice that the debugger will leave the
``fibonacci`` function as soon as you hit a ``co_yield`` statement. You might
find yourself inside some standard library code. After stepping out of the
library code, you will be back in the ``main`` function.

Stepping into a coroutine
-------------------------

If you stop at ``++fib`` and try to step into the generator, you will first
find yourself inside ``operator++``. Stepping into the ``handle.resume()`` will
not work by default.

This is because lldb does not step into functions from the standard library by
default. To make this work, you first need to run ``settings set
target.process.thread.step-avoid-regexp ""``. You can do so from the "Debug
Console" towards the bottom of the screen. With that setting change, you can

> step through ``coroutine_handle::resume`` and into your generator.

Reviewer 2: This is cool. I assume we can get a similar behavior with gcc by `b mangled_function_ame.resume`?

And an another related question/problem/feeling is,  given I don't use lldb, I don't have a lot concrete feeling when reading this. I feel there are much more users for gdb than lldb. So maybe it will can help more people to describe some gdb commands here. Even if saying, due to whatever reasons, this is not supported in gdb for some features.

Author: > I assume we can get a similar behavior with gcc by b mangled_function_ame.resume?

Not sure. Does gdb also avoid stepping into standard library functions? Or might a normal "step into" actually just work out of the box?

> I feel there are much more users for gdb than lldb. So maybe it will can help more people to describe some gdb commands here. Even if saying, due to whatever reasons, this is not supported in gdb for some features.

Agree that we might want to also add more context for gdb.
But I don't use gdb and hence don't how things would work in gdb.

Happy to add paragraphs on gdb, in case you can provide me the corresponding gdb comments here as review comments. Alternatively, we might also merge this PR with a focus on lldb, and then you could add the gdb equivalents in a follow-up PR? Let me know what you prefer 🙂

Reviewer 2: Agreed.

the resume function, devirtualization would hence fail for all coroutines that
have reached their final suspension point.

Interpreting the coroutine frame in optimized builds
----------------------------------------------------

The ``__coro_frame`` variable usually refers to the coroutine frame of an
*in-flight* coroutine. This means, the coroutine is currently executing.
However, the compiler only guarantees the coroutine frame to be in a consistent
state while the coroutine is suspended. As such, the variables inside the
``__coro_frame`` variable might be outdated, in particular when optimizations
are enabled.

Furthermore, when optimizations are enabled, the compiler will layout the
coroutine frame more aggressively. Unused values are optimized out, and the
state will usually contain only the minimal information required to reconstruct
the coroutine's state.

clang / LLVM usually use variables like ``__int_32_0`` to represent this
optimized storage. Those values usually do not directly correspond to variables
in the source code.

For example, when compiling the following program, the compiler creates a
single entry ``__int_32_0`` in the coroutine state. Intuitively, one might

> assume that ``__int_32_0`` represents the value of the local variable ``a``.

Reviewer 2: nit: It may be better to hoist the code example. Since I don't know what  is 'a' when I read this.

          unfiltered = ov_parser.unfiltered_arg

          frame = exe_ctx.GetFrame()
          if not frame.IsValid():
              result.SetError("invalid frame")
              return

          if len(args_array) > 1:
              result.SetError("At most one expression expected")
              return
          elif len(args_array) == 1:
              expr = args_array.GetItemAtIndex(0).GetStringValue(9999)
              coro_hdl = frame.EvaluateExpression(expr)
              if not coro_hdl.error.Success():
                  result.AppendMessage(
                      f'error: expression failed {expr} => {async_root.error}'
                  )
                  result.SetError(f"Expression `{expr}` failed to evaluate")
                  return
              _print_async_bt(coro_hdl, result,
                  curr_idx = 0, start = frame_index, limit = frame_index + count,
                  continuation_paths = continuation_paths)
          else:
              _print_combined_bt(frame, result, unfiltered=unfiltered,
                  curr_idx = 0, start = frame_index, limit = frame_index + count,
                  continuation_paths = continuation_paths)

>   class Coroin-flightCommand(ParsedCommand):

Author: accidentally introduced when I did a global `s/inflight/in-flight`. Fixed
